File: examples/NBomber.Http.Api2.CSharp/SequentialSteps.cs
using System;
using NBomber.CSharp;

namespace NBomber.Http.Api2.CSharp
{
    public class SequentialSteps
    {
        public static void Run()
        {
            var step1 = HttpStep.Create("step 1", ctx =>
                Http.CreateRequest("https://gitter.im")); // defaults to GET without headers

            var step2 = HttpStep.Create("step 2", async ctx =>
            {
                var response = ctx.GetPreviousStepResponse();
                var headers = response.Headers;
                var body = await response.Content.ReadAsStringAsync();
                return Http.CreateRequest("https://gitter.im");
            });
            var scenario = Scenario.Create("test gitter",
                concurrentCopies: 100,
                duration : TimeSpan.FromSeconds(10),
                step1, step2
            );
            NBomberRunner.RegisterScenarios(scenario).RunInConsole();
        }
    }
}

File: examples/NBomber.Http.Api2.FSharp/SequentialSteps.fs
module SequentialSteps

open System
open System.Net.Http
open FSharp.Control.Tasks.V2.ContextInsensitive
open NBomber.FSharp
open NBomber.Http
open NBomber.Http.Api2.FSharp

let run () =

    let step1 =
        Http.request "https://gitter.im"
        |> HttpStep.withRequest "step0"
        
    let step2 = 
        HttpStep.create("step 1", fun _ ->
            { Http.request "https://gitter.im" with
                Headers = [| "Accept", "text/html" |] }
        )

    let step3 = 
        HttpStep.create("step 2", fun context -> task {
            
            let step1Response = context.GetPreviousStepResponse()            
            let headers = step1Response.Headers
                          |> Seq.map (|KeyValue|)
                          |> Seq.map(fun(k,v) -> k, Seq.head v)
                          |> Array.ofSeq
            let! body = step1Response.Content.ReadAsStringAsync()
            
            return { Http.request "https://github.com" with
                        Method = HttpMethod.Post
                        Headers = [| "Accept", "text/html"
                                     "AcceptLanguage", "de-DE" |]
                                  |> Array.append headers
                        Body = new StringContent(body) }
        })

    let scenario = 
        Scenario.create "test gitter and github" [step1; step2; step3]
        |> Scenario.withConcurrentCopies 1
        |> Scenario.withDuration(TimeSpan.FromSeconds 10.)

    NBomberRunner.registerScenarios [scenario]
    |> NBomberRunner.runInConsole